va-combo-box: add option groups functionality

oleksii-morgun opened the pull request:
## Chromatic
<!-- This `combobox-option-groups` is a placeholder for a CI job - it will be updated automatically -->
https://combobox-option-groups--65a6e2ed2314f7b8f98609d8.chromatic.com

---
## Configuring this pull request
- [x] Link to any related issues in the description so they can be cross-referenced.
- [x] Add the appropriate version patch label (`major`, `minor`, `patch`, or `ignore-for-release`).
    - See [How to choose a version number](https://github.com/department-of-veterans-affairs/component-library#how-to-choose-a-version-number) for guidance.
    - Use `ignore-for-release` if files haven't been changed in a component library package. (ie. only Storybook files)
- [ ] DST Only: Increment the `/packages/core` version number if this will be the last PR merged before a release.
- [ ] Complete all sections below.
- [ ] Delete this section once complete

## Description
https://www.figma.com/design/uddjxccqvUyz5FSg1aD5wi/Z---Triage-Group-Flows?node-id=2095-166308&t=N9EvV05ND53hcKNY-0
- adding support for `optgroup` tag that is expected to behave similar to `select` dropdown `optgroup`
- `optgroup` is transformed to a `<li>` in a combobox list
- `optgroup` receives custom distinct styling
- `optgroup` is not selectable or searchable
- child `option` elements of an `optgroup` receive `aria-describedby` of an `optgroup` they are associated with
- child `option` element that matches the search query is displayed along with its parent `optgroup` element
![image](https://github.com/user-attachments/assets/990b1001-a029-48c2-ab00-1f93b2aeb2f6)

![image](https://github.com/user-attachments/assets/0eeda0da-212a-419f-91e5-dab0c2083581)

![2025-02-07 11 29 14](https://github.com/user-attachments/assets/a7901dfd-34cb-49ee-81b7-aaf2382f5d4c)




## QA Checklist
- [x] Component maintains 1:1 parity with design mocks
- [ ] Text is consistent with what's been provided in the mocks
- [x] Component behaves as expected across breakpoints
- [x] Accessibility expert has signed off on code changes (if applicable. If not applicable provide reason why)
- [x] Designer has signed off on changes (if applicable. If not applicable provide reason why)
- [x] Tab order and focus state work as expected
- [x] Changes have been tested against screen readers (if applicable. If not applicable provide reason why)
- [x] New components are covered by e2e tests; updates to existing components are covered by existing test suite
- [ ] Changes have been tested in vets-website using Verdaccio (if applicable. If not applicable provide reason why)

## Screenshots


## Acceptance criteria
- [ ] QA checklist has been completed
- [ ] Screenshots have been attached that cover desktop and mobile screens

## Definition of done
- [x] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

oleksii-morgun commented:
@BobbyBaileyRB has conducted a11y testing on our end as well

Andrew565 commented on packages/storybook/stories/va-combo-box-uswds.stories.jsx:
```diff
@@ -46,6 +46,27 @@ const defaultArgs = {
   ],
 };
 
+const optGroupArgs = {
```
Please move this down to just before the `export const OptGroups` below so that it's more obvious that these args go with that template.

oleksii-morgun replied:
👍🏻

Andrew565 commented on packages/storybook/stories/va-combo-box-uswds.stories.jsx:
```diff
@@ -46,6 +46,27 @@ const defaultArgs = {
+        <option value="Potatoes">Potatoes</option>
+      </optgroup>
+      <option value="sweet-potatoes">Sweet Potatoes</option>
+      <option value="cherry">Cherry</option>
```
I believe most people would be likely to associate "Cherry" and "Clementine" as "Fruits", not "Vegetables".

oleksii-morgun replied:
In this case, `Sweet potatoes`, `Cherry`, and `Clementine` are intended to fall in neither of those two categories to demonstrate that options can be either nested or not. Maybe I just need to pick examples that are distinctively different from those categories. I also wanted to pick some non nested values outside of an `<optgroup>` that have matching segments, to demonstrate filter functionality too

Andrew565 commented on packages/storybook/stories/va-combo-box-uswds.stories.jsx:
```diff
@@ -115,3 +136,8 @@ WithMessageAriaDescribedBy.args = {
   messageAriaDescribedby: 'This is example aria message',
 };
+
+export const OptGroups = Template.bind({});
```
Please expand the name of this template variable to "OptionGroups", as Storybook will automatically expand this to "Option Groups". "Opt Groups" is not "obvious" enough, in my opinion.

Andrew565 commented on packages/web-components/src/components/va-combo-box/va-combo-box-library.js:
```diff
@@ -368,6 +370,8 @@ const noop = () => {};
     return new RegExp(find, 'i');
   };
 
+  const isDataOptGroup = option => option.getAttribute('data-optgroup') !== null;
```
This should probably be expanded to include `&& option.getAttribute('data-optgroup') !== false`.

oleksii-morgun replied:
now that I am giving this another look, it probably should just be `option.getAttribute('data-optgroup') === 'true';`

Andrew565 commented on packages/web-components/src/components/va-combo-box/va-combo-box-library.js:
```diff
@@ -411,37 +416,77 @@ const noop = () => {};
         }
-        options.push(optionEl);
+
+        if (!inputValue){
+          options.push(optionEl);
+          continue;
+        }
+
+        if (
+          inputValue &&
+          regex.test(optionEl.text) &&
+          optionEl.getAttribute('data-optgroup') !== 'true' 
+        ) {
+          if (
+            optionEl.getAttribute('data-optgroup-option') === 'true' &&
+            parentOptGroupId !== optionEl.getAttribute('aria-describedby')
+          ) {
+            parentOptGroupId = optionEl.getAttribute('aria-describedby');
+            const parentOptgroupEl = selectEl.querySelector(
+              '#' + parentOptGroupId,
+            );
+            options.push(parentOptgroupEl);
+          }
+          options.push(optionEl);
+        }
```
Can you add some comments to help make it more obvious what this code is doing?

micahchiang reviewed (commented):
Hey @oleksii-morgun! Thanks for working on this some more, we really appreciate the contribution. When you have a spare moment, would you mind leaving some review comments around the stuff going on `va-combo-box-library.js` and `va-combo-box.tsx`, just so we have a little more context? Thanks!

LWWright7 reviewed (changes requested):
The [current deployment](https://design.va.gov/storybook/?path=/docs/uswds-va-combo-box--docs) of Storybook shows that when a user selects the dropdown arrow to look at the available options, the focus ring moves to the first item in the list and is being highlighted in the focus default yellow `#FACE00`
![Screen Shot 2025-02-07 at 10 40 11 AM](https://github.com/user-attachments/assets/7160b34b-12a1-4e09-a9f1-9c63c66078b9)

However, with this [updated link](https://combobox-option-groups--65a6e2ed2314f7b8f98609d8.chromatic.com/) it appears that the focus color has changed to black:
![Screen Shot 2025-02-07 at 10 30 35 AM](https://github.com/user-attachments/assets/64e513f7-4e7c-4d41-a96e-949ae7568a3c)

Correct me if I'm wrong, but I believe this should be in the focus yellow color mentioned above.

I'm guessing that the main reason for this is because, when an item from the list is selected and the user reengages the dropdown button, the highlighted option is shown in dark blue and if the black default color were used, I'm not sure it would pass  <kbd>a11y</kbd>  due to lack of contrast with the dark colors used, as shown here:
![Screen Shot 2025-02-07 at 10 59 41 AM](https://github.com/user-attachments/assets/3dc39f88-9982-467f-b5c6-38078811fb0f)

But perhaps @rsmithadhoc could provide some more insight on this particular issue?

danbrady commented:
Great work, Alex! The design looks great. One minor suggestion, can we add a bit more left padding on the items within the optgroups? If you can't see the optgroup header, extra padding will help differentiate the items that are and are not part of an optgroup. It's currently 16px, but I'm leaning towards 24px.

<img width="400" alt="Screenshot 2025-02-07 at 12 07 53 PM" src="https://github.com/user-attachments/assets/40469b78-edd9-4d12-82e7-06c4da9a50fa" /><br>

<img width="400" alt="Screenshot 2025-02-07 at 12 08 24 PM" src="https://github.com/user-attachments/assets/d6838440-192e-4775-8fb3-3017f1ba1199" />

Alternatively, maybe the last item in an optgroup gets a thicker border?


<img width="400" alt="Screenshot 2025-02-07 at 12 20 39 PM" src="https://github.com/user-attachments/assets/9e19c47d-9542-4701-8638-8fa2e6e74b13" />

rsmithadhoc commented:
> The [current deployment](https://design.va.gov/storybook/?path=/docs/uswds-va-combo-box--docs) of Storybook shows that when a user selects the dropdown arrow to look at the available options, the focus ring moves to the first item in the list and is being highlighted in the focus default yellow `#FACE00` ![Screen Shot 2025-02-07 at 10 40 11 AM](https://private-user-images.githubusercontent.com/146852077/411001537-7160b34b-12a1-4e09-a9f1-9c63c66078b9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyMDYxNTMsIm5iZiI6MTczOTIwNTg1MywicGF0aCI6Ii8xNDY4NTIwNzcvNDExMDAxNTM3LTcxNjBiMzRiLTEyYTEtNGUwOS1hOWYxLTljNjNjNjYwNzhiOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxMFQxNjQ0MTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MWU4NjhkMDgwNTMzMDBiZGIyOTEwMDM2NjBhNzYyMDUwODkzZjg2MjQ1ZTg1ODAwNWNiYjMzNjc0NzViZTY5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.1UT7NVPx3KYqo8Zo2CyyFNXDYy6dzOtpjACVPF0-EAo)
> 
> However, with this [updated link](https://combobox-option-groups--65a6e2ed2314f7b8f98609d8.chromatic.com/) it appears that the focus color has changed to black: ![Screen Shot 2025-02-07 at 10 30 35 AM](https://private-user-images.githubusercontent.com/146852077/411002318-64e513f7-4e7c-4d41-a96e-949ae7568a3c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyMDYxNTMsIm5iZiI6MTczOTIwNTg1MywicGF0aCI6Ii8xNDY4NTIwNzcvNDExMDAyMzE4LTY0ZTUxM2Y3LTRlN2MtNGQ0MS1hOTZlLTk0OWFlNzU2OGEzYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxMFQxNjQ0MTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MDBmMTFmNjljOTJiNGQ2OWNjOWQwYjAxZmRhMDNiOTg1MWFhZGM4MTUxNmRiZWI5NjZhZjg0Y2YwOTg1NjAyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Xaj66aRQjpgRoDUAdGkYfFhbj6_Cp1oI9fg_RH60OSY)
> 
> Correct me if I'm wrong, but I believe this should be in the focus yellow color mentioned above.
> 
> I'm guessing that the main reason for this is because, when an item from the list is selected and the user reengages the dropdown button, the highlighted option is shown in dark blue and if the black default color were used, I'm not sure it would pass a11y due to lack of contrast with the dark colors used, as shown here: ![Screen Shot 2025-02-07 at 10 59 41 AM](https://private-user-images.githubusercontent.com/146852077/411006328-3dc39f88-9982-467f-b5c6-38078811fb0f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyMDYxNTMsIm5iZiI6MTczOTIwNTg1MywicGF0aCI6Ii8xNDY4NTIwNzcvNDExMDA2MzI4LTNkYzM5Zjg4LTk5ODItNDY3Zi1iNWM2LTM4MDc4ODExZmIwZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxMFQxNjQ0MTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05MTdkZDliZGY2MmQ1NDFiYjFmOGQxY2E5OWJlOThlYTNiMTRiMjI3Yjk2MDBmMzY2YWU2Y2U1NmU1OTE2MWU2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.BL49m3wq_UEffZowr7MLq56ZuGnTL5Au7oqOayf_4pk)
> 
> But perhaps @rsmithadhoc could provide some more insight on this particular issue?

@LWWright7 
I'm not completely sure on this one, so I'm open to other opinions, but here are my thoughts:

- The version in the PR matches USWDS. 
- The black outline shows when you are not focused on the options, but focus is on the input (and it has the yellow outline). The black rectangle acts as a current selection indicator, rather than a focus indicator.
- The same holds for the black and dark blue, nothing is focused at that point. In that case, I would say the boundary of that rectangle (black) should have a 3:1 contrast ratio to the adjacent background (white), which it does.
  - The black outline against the dark blue rectangle is also just above a 3:1 ratio. I'm not sure if that would violate anything if it were lower, but I think it passes the minimum threshold if that were the case.
- When navigating the options, it shows the yellow focus indicator correctly.
- Having the double yellow might confuse when the option itself is not currently being focused. 

For those reasons, I think the current version should be fine.

oleksii-morgun commented:
> Hey @oleksii-morgun! Thanks for working on this some more, we really appreciate the contribution. When you have a spare moment, would you mind leaving some review comments around the stuff going on `va-combo-box-library.js` and `va-combo-box.tsx`, just so we have a little more context? Thanks!

definitely. `va-combo-box-library.js` is mostly inherited from USWDS with an exception of few tweaks and what is in this PR to support `optgroup` - adding those comments in here .

oleksii-morgun commented:
re: https://github.com/department-of-veterans-affairs/component-library/pull/1478#pullrequestreview-2602352396

@LWWright7 here is a reference to a thread where this was brought up by @BobbyBaileyRB and it seems to align with @rsmithadhoc point. But I am open to your collective guidance here

oleksii-morgun commented:
re: https://github.com/department-of-veterans-affairs/component-library/pull/1478#issuecomment-2643540925

@JoelCalumpong any concerns from Design perspective? I know we are considering this primarily for recipient list where we are trying to save as much real estate as possible.

oleksii-morgun commented:
I pushed updates addressing some concerns and including feedback followup from all your input. Only item pending is the padding confirmation by our design team @JoelCalumpong. Other than that, should be ready for re-review when y'all have. a chance please
